include/seqan3/range/container/bitcompressed_vector.hpp

```diff
@@ -0,0 +1,1114 @@
+// FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+// LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+// OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+// DAMAGE.
+//
+// ============================================================================
+
+/*!\file
+ * \brief Provides seqan3::bitcompressed_vector.
+ * \author Hannes Hauswedell <hannes.hauswedell AT fu-berlin.de>
+ */
+
+#pragma once
+
+#include <type_traits>
+
+#include <range/v3/iterator_range.hpp>
+
+#include <sdsl/int_vector.hpp>
+
+#include <seqan3/core/concept/cereal.hpp>
+#include <seqan3/core/metafunction/range.hpp>
+#include <seqan3/range/container/concept.hpp>
+#include <seqan3/range/detail/random_access_iterator.hpp>
+#include <seqan3/range/view/to_rank.hpp>
+#include <seqan3/range/view/convert.hpp>
+#include <seqan3/std/concept/iterator.hpp>
+
+namespace seqan3
+{
+
+/*!\brief A space-optimised version of std::vector that compresses multiple letters into a single byte.
+ * \tparam alphabet_type The value type of the container, must satisfy seqan3::alphabet_concept and not be `const` or
+ *                       `&`.
+ * \implements seqan3::reservable_container_concept
+ * \implements seqan3::cerealisable_concept
+ * \ingroup container
+ *
+ * This class template behaves just like std::vector<alphabet_type> but has an internal representation where
+ * multiple values are packed into a single byte/word to save space, e.g. bitcompressed_vector<seqan3::dna4> uses a
+ * quarter of of the memory that std::vector<seqan3::dna4> uses, because a single seqan3::dna4 letter can be represented
+ * in two bits (instead of 8 which is the lower bound for a single object in C++.
+ *
+ * The disadvantages are slightly slower operations and unsafety towards parallel writes to adjacent positions
+ * in the bitcompressed_vector.
+ *
+ *
+ * ### Example
+ *
+ * ```cpp
+ * std::vector<dna4>          v0{"ACGT"_dna4}; // occupies 4 bytes in memory
+ * bitcompressed_vector<dna4> v1{"ACGT"_dna4}; // occupies 1 byte in memory
+ * ```
+ *
+ * ### Thread safety
+ *
+ * This container provides no thread-safety beyond the promise given also by the STL that all
+ * calls to `const` member function are safe from multiple threads (as long as no thread calls
+ * a non-`const` member function at the same time).
+ *
+ * An important difference to std::vector is that calling `vec[0] = value;` and `vec[1] = value2;` from two different
+ * threads at the same time **is not safe** and will likely lead to corruption.
+ */
+template <typename alphabet_type>
+//!\cond
+    requires alphabet_concept<alphabet_type> && std::is_same_v<alphabet_type, remove_cvref_t<alphabet_type>>
```

Comment:
why not the first expression as intermediate concept declaration?

Reply:
It wasn't clear if we were using them at the point of writing.